src/AspNet.Security.OAuth.Automatic/AutomaticAuthenticationHandler.cs

```diff
@@ -0,0 +1,78 @@
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Automatic {
+    internal class AutomaticAuthenticationHandler : OAuthHandler<AutomaticAuthenticationOptions> {
+        public AutomaticAuthenticationHandler([NotNull] HttpClient httpClient)
+            : base(httpClient) {}
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens) {
+            var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode) {
+                Logger.LogError("An error occurred when retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred when retrieving the user profile.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, AutomaticHelper.GetIdentifier(payload),
+                    Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.GivenName, AutomaticHelper.GetGivenName(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Surname, AutomaticHelper.GetFamilyName(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Name, AutomaticHelper.GetLogin(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Email, AutomaticHelper.GetEmail(payload), Options.ClaimsIssuer)
+                .AddOptionalClaim(ClaimTypes.Uri, AutomaticHelper.GetUrl(payload), Options.ClaimsIssuer);
+
+            var principal = new ClaimsPrincipal(identity);
+            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
+
+            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
+            await Options.Events.CreatingTicket(context);
+
+            return context.Ticket;
+        }
+
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri) {
+            var scopes = FormatScope();
+
+            var queryStrings = new Dictionary<string, string> {
```

Comment:
No, you have to create an app then specify the redirect uri there. I
haven't tried to send one but it's not mentioned in the docs at all. I'll
give it a shot anyway.

On Friday, July 1, 2016, Kévin Chalet <notifications@github.com> wrote:

> In src/AspNet.Security.OAuth.Automatic/AutomaticAuthenticationHandler.cs
> <https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/pull/104#discussion_r69347518>
> :
>
> > +                .AddOptionalClaim(ClaimTypes.Email, AutomaticHelper.GetEmail(payload), Options.ClaimsIssuer)
> > +                .AddOptionalClaim(ClaimTypes.Uri, AutomaticHelper.GetUrl(payload), Options.ClaimsIssuer);
> > +
> > +            var principal = new ClaimsPrincipal(identity);
> > +            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
> > +
> > +            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
> > +            await Options.Events.CreatingTicket(context);
> > +
> > +            return context.Ticket;
> > +        }
> > +
> > +        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri) {
> > +            var scopes = FormatScope();
> > +
> > +            var queryStrings = new Dictionary<string, string> {
>
> No redirect_uri? That's the first time I see a provider that doesn't
> support (or requires) this parameter. What happens if you send it anyway?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/pull/104/files/b232e644da5dc1bd08cb07eac025bf7ff533f1b1#r69347518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABDhmwcaFAzutCyPYjNqarS4XiLkWYF1ks5qRW6bgaJpZM4JC3Om>
> .
>